app/scripts/controllers/plugins.js

```diff
@@ -363,45 +484,107 @@ class PluginsController extends EventEmitter {
+    return eventEmitter.eventNames().reduce((acc, eventName) => {
+
+      if (pluginRestrictedMethodDescriptions[eventName]) {
+
+        // keep a reference to the emitter so we can add and remove listeners
+        acc[eventName] = {
+          eventEmitter,
+        }
       }
-    })
+      return acc
+    }, {})
   }
 
-  generateMetaMaskListenerMethodsMap () {
-    return [
-      ...this._eventEmitterToListenerMap(this._txController),
-      ...this._eventEmitterToListenerMap(this._networkController),
-      ...this._eventEmitterToListenerMap(this._blockTracker),
-    ].reduce((acc, methodMap) => ({ ...acc, ...methodMap }))
+  /**
+   * Used in constructor to generate a map of event names to EventEmitters for
+   * adding and removing event listeners.
+   * See _eventEmitterToListenerMap for implementation details.
+   *
+   * @param {Array<EventEmitter>} eventEmitters - The EventEmitters to get event names from.
+   */
+  _generateMetaMaskListenerMethodsMap (eventEmitters) {
+    return eventEmitters.reduce((acc, emitter) => {
+      return { ...acc, ...this._eventEmitterToListenerMap(emitter)}
+    }, {})
   }
 
-  _createMetaMaskEventListener (approvedPermissions) {
-    const listenerMethodsMap = this.generateMetaMaskListenerMethodsMap()
-    const approvedListenerMethodsMap = {}
+  /**
+   * Creates a function for the given plugin to listen to MetaMask events.
+   *
+   * @param {string} pluginName - The name of the plugin.
+   * @param {Array<string>} approvedPermissions - The names of the plugin's approved permissions.
+   */
+  _createMetaMaskEventListener (pluginName, approvedPermissions) {
+
+    const approvedListenerMethods = {}
     approvedPermissions.forEach(approvedPermission => {
-      if (listenerMethodsMap[approvedPermission]) {
-        approvedListenerMethodsMap[approvedPermission] = listenerMethodsMap[approvedPermission]
+      if (this._listenerMethods[approvedPermission]) {
+        approvedListenerMethods[approvedPermission] = this._listenerMethods[approvedPermission]
       }
     })
 
+    // keep track of listeners for later removal
+    const registeredListenersMap = new Map()
+    this.metamaskEventListeners.set(pluginName, registeredListenersMap)
+
     return (eventName, cb) => {
-      approvedListenerMethodsMap[eventName](cb)
+
+      // plugin may have been removed and not yet garbage collected
+      if (!this.metamaskEventListeners.has(pluginName)) {
+        return
+      }
+
+      // throw error if method is unknown or unpermitted
+      if (!approvedListenerMethods[eventName]) {
+        if (!this._listenerMethods[eventName]) {
+          throw new Error('Unknown event name.')
+        } else {
+          throw new Error('Unpermitted event name.')
+        }
+      }
+
+      // remove any existing listener
+      if (registeredListenersMap.has(eventName)) {
+        // call the stored removal function, it will be replaced later
+        registeredListenersMap.get(eventName)()
```

Comment:
We probably want to rename this listeners map to indicate it's a map of removal functions.